#
# Propagate errors returned by pam_acct_mgmt() to the upper layer,
# so that user can be informed why the call really failed.
# Without these changes, all errors are mapped to PAM_ACCT_EXPIRED.
#
# This issue is tracked by upstream via
# https://bugzilla.mindrot.org/show_bug.cgi?id=3210
#
# The patch can be removed in case the above bugzilla gets fixed and the fix
# addresses the problem in desired way. 
#
--- old/auth-pam.c	2025-03-06 11:44:42.907013110 +0100
+++ new/auth-pam.c	2025-03-06 11:44:43.475514830 +0100
@@ -418,10 +418,10 @@ do_pam_auth(struct pam_ctxt *ctxt)
 	if (sshpam_err != PAM_SUCCESS)
 		goto auth_fail;
 
-	if (!do_pam_account()) {
-		sshpam_err = PAM_ACCT_EXPIRED;
+	sshpam_err = do_pam_account();
+	if (sshpam_err != PAM_SUCCESS)
 		goto auth_fail;
-	}
+
 	if (sshpam_authctxt->force_pwchange) {
 		if ((buffer = sshbuf_new()) == NULL)
 			fatal("%s: sshbuf_new failed", __func__);
@@ -1078,26 +1078,32 @@ finish_pam(void)
 u_int
 do_pam_account(void)
 {
+	static int sshpam_err_saved = PAM_SUCCESS;
+
 	debug("%s: called", __func__);
-	if (sshpam_account_status != -1)
-		return (sshpam_account_status);
+	if (sshpam_account_status != -1) {
+		return (sshpam_err_saved);
+	}
 
 	expose_authinfo(__func__);
 
 	sshpam_err = pam_acct_mgmt(sshpam_handle, 0);
+	sshpam_err_saved = sshpam_err;
 	debug3("PAM: %s pam_acct_mgmt = %d (%s)", __func__, sshpam_err,
 	    pam_strerror(sshpam_handle, sshpam_err));
 
 	if (sshpam_err != PAM_SUCCESS && sshpam_err != PAM_NEW_AUTHTOK_REQD) {
 		sshpam_account_status = 0;
-		return (sshpam_account_status);
+		return (sshpam_err);
 	}
 
-	if (sshpam_err == PAM_NEW_AUTHTOK_REQD)
+	if (sshpam_err == PAM_NEW_AUTHTOK_REQD) {
 		sshpam_password_change_required(1);
+		sshpam_err = sshpam_err_saved = PAM_SUCCESS;
+	}
 
 	sshpam_account_status = 1;
-	return (sshpam_account_status);
+	return (sshpam_err);
 }
 
 void
--- old/auth2.c	2025-03-06 11:44:40.592102180 +0100
+++ new/auth2.c	2025-03-06 11:44:43.475998160 +0100
@@ -421,7 +421,7 @@ userauth_finish(struct ssh *ssh, int aut
 		 * perform account check for this service.
 		 */
 		if (options.use_pam && options.pam_service_per_authmethod &&
-		    !mm_do_pam_account()) {
+		    mm_do_pam_account() != PAM_SUCCESS) {
 			/* if PAM returned a message, send it to the user */
 			if (sshbuf_len(loginmsg) > 0) {
 				if ((r = sshbuf_put(loginmsg, "\0", 1)) != 0)
@@ -468,7 +468,7 @@ userauth_finish(struct ssh *ssh, int aut
 #else
 	if (options.use_pam && authenticated) {
 #endif /* PAM_ENHANCEMENT */
-		int r, success = mm_do_pam_account();
+		int r, success = mm_do_pam_account() == PAM_SUCCESS;
 
 		/* If PAM returned a message, send it to the user. */
 		if (sshbuf_len(loginmsg) > 0) {
--- old/monitor.c	2025-03-06 11:44:41.735677570 +0100
+++ new/monitor.c	2025-03-06 11:44:43.476687545 +0100
@@ -1136,7 +1136,7 @@ mm_answer_pam_account(struct ssh *ssh, i
 
 	mm_request_send(sock, MONITOR_ANS_PAM_ACCOUNT, m);
 
-	return (ret);
+	return (ret == PAM_SUCCESS);
 }
 
 static void *sshpam_ctxt, *sshpam_authok;
